xds/src/test/java/io/grpc/xds/ClientXdsClientDataTest.java

```diff
@@ -773,245 +803,197 @@ public void parseCluster_ringHashLbPolicy_invalidRingSizeConfig_tooLargeRingSize
-        .isEqualTo("filter envoy.http_connection_manager expected to have typed_config");
+    thrown.expect(ResourceInvalidException.class);
+    thrown.expectMessage(
+        "FilterChain filter-chain-foo contains filter envoy.http_connection_manager "
+            + "without typed_config");
+    ClientXdsClient.parseServerSideListener(listener, null, true);
   }
 
   @Test
-  public void parseServerSideListener_wrongTypeUrl() {
+  public void parseServerSideListener_wrongTypeUrl() throws ResourceInvalidException {
     Filter filter =
         Filter.newBuilder()
             .setName("envoy.http_connection_manager")
             .setTypedConfig(Any.newBuilder().setTypeUrl("badTypeUrl"))
             .build();
-    FilterChain filterChain = buildFilterChain(filter);
+    FilterChain filterChain = buildFilterChain("filter-chain-foo", filter);
     Listener listener =
         Listener.newBuilder()
             .setName("listener1")
             .setTrafficDirection(TrafficDirection.INBOUND)
             .addFilterChains(filterChain)
+            .setAddress(
+                Address.newBuilder()
+                    .setSocketAddress(
+                        SocketAddress.newBuilder().setAddress("10.0.1.2").setPortValue(8080)))
             .build();
-    StructOrError<io.grpc.xds.EnvoyServerProtoData.Listener> struct =
-        ClientXdsClient.parseServerSideListener(listener, null);
-    assertThat(struct.getErrorDetail())
-        .isEqualTo(
-            "filter envoy.http_connection_manager with unsupported typed_config type:badTypeUrl");
+    thrown.expect(ResourceInvalidException.class);
+    thrown.expectMessage(
+        "FilterChain filter-chain-foo contains filter envoy.http_connection_manager with "
+            + "unsupported typed_config type badTypeUrl");
+    ClientXdsClient.parseServerSideListener(listener, null, true);
   }
 
   @Test
-  public void parseServerSideListener_duplicateHttpFilter() {
+  public void parseServerSideListener_duplicateHttpFilter() throws ResourceInvalidException {
     Filter filter =
-        buildHttpConnectionManager(
-            "envoy.http_connection_manager",
+        buildHttpConnectionManagerFilter(
             HttpFilter.newBuilder().setName("hf").setIsOptional(true).build(),
             HttpFilter.newBuilder().setName("hf").setIsOptional(true).build());
-    FilterChain filterChain = buildFilterChain(filter);
-    Listener listener =
-        Listener.newBuilder()
-            .setName("listener1")
-            .setTrafficDirection(TrafficDirection.INBOUND)
-            .addFilterChains(filterChain)
-            .build();
-    StructOrError<io.grpc.xds.EnvoyServerProtoData.Listener> struct =
-        ClientXdsClient.parseServerSideListener(listener, null);
-    assertThat(struct.getErrorDetail())
-        .isEqualTo("http-connection-manager has non-unique http-filter name:hf");
-  }
-
-  @Test
-  public void parseServerSideListener_configDiscoveryHttpFilter() {
```

Comment:
Is this being tested elsewhere?

Reply:
This is one kind of "HttpFilter with unsupported config" test.

```
oneof config_type {
   google.protobuf.Any typed_config

   config.core.v3.ExtensionConfigSource config_discovery
}
```
The implementation only looks up `typed_config` and only cares if `type_config` is present. It's a `oneof`. As long as `typed_config` is missing while it is required, it should fail. Testing on things unrelated is redundant (more strictly speaking, is incorrect as it makes tests fragile). Tests should only _test right things_.

See `parseHttpFilter_unsupportedButOptional()` and `parseHttpFilter_unsupportedAndRequired()` for such.

Reply:
Note I've cleaned up tests for covering validations in parsing server side Listener a bit. Mostly narrow down what each test method is trying to verify to just the corresponding implementation. For example, if the test is trying to verify the validation logic of parsing an HttpConnectionManager, we don't bother creating a Listener containing a FilterChain containing an HttpConnectionManager and verify the correctness of entire parseListener() method. This helps more evenly partition the test space, easier for identifying bugs if there is any and simpler for maintenance.

/cc @sanjaypujare